### PR TITLE
[StandardToHandshake] Remove intermediate conversion ops

### LINKED ISSUE #3686: [StandardToHandshake] Remove intermediate conversion ops
Currently, the `StandardToHandshake` conversion uses several intermediate ops to aid in the lowering pass. Usually, intermediate ops may aid in reducing conversion pass complexity, through splitting up conversion into several smaller passes.
However, the `StartOp, EndOp, TerminatorOp`s of the Handshake dialect are (as far as I am aware of) only used during the `StandardToHandshake` pass, and never live beyond them. Over time, we've experienced a bunch of issues with respect to these. More generally, these issues are linked to maintaining intermediate state in the IR **as it is being rewritten** - which has shown not to be a stable approach to writing conversion passes.

As far as I can tell, most use of the three aforementioned ops have been removed, and I expect it to be a fairly minor change to get rid of the remaining ones. Any load-bearing that these ops still perform should be moved to an in-memory intermediate state object maintained during the conversion pass.

### PR DESCRIPTION
**Pull request by RamirezLucas**
This commit removes the `StartOp`, `EndOp`, and `TerminatorOp` from the entire codebase, since they were used exclusively as intermediate operations during `std-to-handshake` conversion and have caused issues in the past. `StartOp`'s purpose is achieved through a new field (`startCtrl`) in the `HandshakeLowering` class, while `TerminatorOp` and `EndOp`'s purpose is obtained by keeping `cf`-level terminators in the IR for longer than before.

Unit tests on the conversion pass are modified manually (as to yield a smaller diff, since rerunning `FileCheck` would change SSA value names and modify almost all lines) to reflect the fact that an additional `MergeOp` is created by the pass for the control-only network in the entry block of each function (this has no impact on circuit behavior, and can be canonicalized away).

The commit also includes some more detailed documentation for the merge-insertion step of the conversion pass.

Fixes #3686.

### CODE REVIEW

**Review by Dinistro — APPROVED** [review]
LGTM! Thank you very much for pushing on this. 

I suggest to wait for the review of @mortbopet before merging.

**Review by mortbopet — APPROVED** [review]
Brilliant, thanks!

**Comment by RamirezLucas** [comment]
Thanks both for the very quick reviews!

**Pull request merged**
